### PR TITLE
Weapon drop callback + Fix for weapons still dropping

### LINKED ISSUE #613: Executing someone drops weapons in Infection
Steps to reproduce:
1) Be Infected
2) Execute a Survivor
3) Get gun
4) Kill every Survivor

### PR DESCRIPTION
**Pull request by NoCatt**
As of making this PR its untested.

This adds a callback for when a gun gets dropped, this is just an implementation of the already existing CodeCallback.
It at the same time fixes https://github.com/R2Northstar/NorthstarMods/issues/613 and is a better #617 , it removed the callback for deleting the dropped gun.

Closes #613

### CODE REVIEW

**Comment by NoCatt** [comment]
okay so I tested the code callback now and the flag, and both work havent tested in specifically Infection though

File: Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut [hunk 1]
```diff
@@ -31,7 +31,7 @@ void function SvLoadoutsMP_Init()
 	// most of these are fairly insecure right now, could break pdata if called maliciously, need fixing eventually
 	RegisterSignal( "EndUpdateCachedLoadouts" )
 	RegisterSignal( "GracePeriodDone" ) // temp to get weapons\_weapon_utility.nut:2271 to behave
-	
+
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut:34]: Accidental whitespace change?

File: Northstar.CustomServers/mod/scripts/vscripts/mp/_codecallbacks.gnut [hunk 2]
```diff
@@ -43,6 +45,7 @@ struct
 	]
 
 	table<entity, AccumulatedDamageData> playerAccumulatedDamageData
+	array<void functionref( entity )> WeaponDroppedCallbacks
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/_codecallbacks.gnut:48]:
```suggestion
	array< void functionref( entity ) > WeaponDroppedCallbacks
```

File: Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut [hunk 3]
```diff
@@ -62,20 +60,10 @@ void function SetLoadoutGracePeriodEnabled( bool enabled )
...
-		thread DelayDestroyDroppedWeapon( victim.GetActiveWeapon() )
-}
-
-void function DelayDestroyDroppedWeapon( entity weapon )
-{
-	WaitEndFrame()
-	if ( IsValid( weapon ) )
-		weapon.Destroy()
+	if( !enabled )
+		FlagSet("WeaponDropsAllowed")
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut:64]: `FlagSet` basically sets the flag to true, right?  Or does it invert it based on what it was set to in `FlagInit`?

Either way, is this logic correct? `SetWeaponDropsEnabled( true )` would currently do `FlagClear`

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Question and some formatting

File: Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut (hunk 3 above)
Comment (NoCatt) [reply Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut:64]: FlagSet sets it to true yes, also yeah now thinking abou it I shouldnt invert the enable anyways that just makes it worse to read, I will fix that in a sec. ( it was late for me too ^^ )

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good, haven't tested yet

**Review by EladNLG — APPROVED** [review]
Looked at: everything
code is fine, diffs are hard to read but git hates you ig.

File: Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut [hunk 4]
```diff
@@ -62,20 +60,10 @@ void function SetLoadoutGracePeriodEnabled( bool enabled )
...
-		thread DelayDestroyDroppedWeapon( victim.GetActiveWeapon() )
-}
-
-void function DelayDestroyDroppedWeapon( entity weapon )
-{
-	WaitEndFrame()
-	if ( IsValid( weapon ) )
-		weapon.Destroy()
+	if( enabled )
+		FlagSet("WeaponDropsAllowed")
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut:64]:
```suggestion
		FlagSet( "WeaponDropsAllowed" )
```

File: Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut [hunk 5]
```diff
@@ -62,20 +60,10 @@ void function SetLoadoutGracePeriodEnabled( bool enabled )
...
+	else 
+		FlagClear("WeaponDropsAllowed")
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/_loadouts_mp.gnut:66]:
```suggestion
		FlagClear( "WeaponDropsAllowed" )
```

File: Northstar.CustomServers/mod/scripts/vscripts/mp/_codecallbacks.gnut [hunk 6]
```diff
@@ -1030,4 +1033,25 @@ void function CodeCallback_OnServerAnimEvent( entity ent, string eventName )
...
 }
+
+void function AddCallback_OnWeaponDropped( void functionref( entity ) callback ) 
+{
+	file.weaponDroppedCallbacks.append( callback )
+}
+
+void function CodeCallback_WeaponDropped( entity weapon )
+{
+	//shamelessly taken form SP
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/_codecallbacks.gnut:1044]:
```suggestion
	// shamelessly taken form SP
```

File: Northstar.CustomServers/mod/scripts/vscripts/mp/_codecallbacks.gnut [hunk 7]
```diff
@@ -1030,4 +1033,25 @@ void function CodeCallback_OnServerAnimEvent( entity ent, string eventName )
...
+	if ( !IsValid( weapon ) )
+		return
+	//Might look a bit hacky to put it here, but thats how SP does it
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/_codecallbacks.gnut:1047]:
```suggestion

	// Might look a bit hacky to put it here, but thats how SP does it
```

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Formatting stuff

**Review by ASpoonPlaysGames — APPROVED** [review]
Updating review because for some reason its still blocking?
Anyway yeah code looks good

**Comment by Bobbyperson12** [comment]
Tested on infection with latest northstar version. Can confirm that this works as intended, executions no longer bypass the weapon drop flag.

**Pull request merged**